dart/lib/src/http_client/failed_request_client.dart

```diff
@@ -0,0 +1,239 @@
+import 'package:http/http.dart';
+import '../protocol.dart';
+import '../hub.dart';
+import '../hub_adapter.dart';
+import '../throwable_mechanism.dart';
+import 'sentry_http_client.dart';
+
+/// A [http](https://pub.dev/packages/http)-package compatible HTTP client
+/// which records events for failed requests.
+///
+/// Configured with default values, this captures requests which throw an
+/// exception.
+/// This can be for example for the following reasons:
+/// - In an browser environment this can be requests which fail because of CORS.
+/// - In an mobile or desktop application this can be requests which failed
+///   because the connection was interrupted.
+///
+/// Additionally you can configure specific HTTP response codes to be considered
+/// as a failed request. In the following example, the status codes 404 and 500
+/// are considered a failed request.
+///
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+///
+/// var client = FailedRequestClient(
+///   failedRequestStatusCodes: [SentryStatusCode.range(400, 404), SentryStatusCode(500)]
+/// );
+/// ```
+///
+/// Remarks:
+/// If this client is used as a wrapper, a call to close also closes the
+/// given client.
+///
+/// The `FailedRequestClient` can be used as a standalone client like this:
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+///
+/// var client = FailedRequestClient();
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+///
+/// The `FailedRequestClient` can also be used as a wrapper for your own
+/// HTTP [Client](https://pub.dev/documentation/http/latest/http/Client-class.html):
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+/// import 'package:http/http.dart' as http;
+///
+/// final myClient = http.Client();
+///
+/// var client = FailedRequestClient(client: myClient);
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+class FailedRequestClient extends BaseClient {
+  FailedRequestClient({
+    this.maxRequestBodySize = MaxRequestBodySize.small,
+    this.failedRequestStatusCodes = const [],
```

Comment:
maybe we have a sort of default value here? otherwise, it won't capture anything ootb.
eg 400...599

Reply:
Hm, I'm not sure.

The [OpenApiGenerator](https://openapi-generator.tech/docs/generators/dart/) throws exceptions if the sever responds with 4xx to 5xx status codes. I'm not quite sure if we could deduplicate those errors.
Then there's also [chopper](https://pub.dev/packages/chopper) and [retrofit](https://pub.dev/packages/retrofit). I would have to look up, how they handle those status code.

My concern is that we would generate way to much noise by including them by default.

We could provide however a couple named constructors for common scenarios:
- server failed (5xx)
- client sent wrong request (4xx)
- all (4xx to 5xx)

Reply:
I see, I think the named ctors would be nice, but not adding anything by default sounds suboptimal too, you're setting up the client already, which means you already want to capture events, the granularity could be configured but should offer something by default IMO.

iirc retrofit only throws if there was a network error, parsing error, or something like that, but if an API returns 400 to 599, you can still read the response, so it does not throw, I believe the `http` Dart is the same, if the Library throws when it returns 4xx or 5xx, this would be almost useless if you send a body response, you'd not be able to read it.

the way how it is, I believe it'd never capture anything when returning 4xx or 5xx by default